Add SDL 7.0 ruleset to code analysis and fix FxCop issues

scottbommarito opened the pull request:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems?id=244727&_a=edit

Everything should be resolved now!

@maartenba @skofman1 @shishirx34 @ryuyu @blowdart

blowdart commented:
Wait, you still have webforms? :)

The user is authenticated at this point? Set the ViewStateUserKey to be the username, or the session ID, or combine them. It's still necessary, because CSRF.

shishirx34 commented on src/NuGet Backend With SDL.ruleset:
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
may be get rid of the spaces in the file name?

maartenba commented on src/NuGetGallery/Areas/Admin/DynamicData/Default.aspx.cs:
```diff
@@ -8,6 +8,12 @@ namespace NuGetGallery.Areas.Admin.DynamicData
+        protected override void OnInit(EventArgs e)
+        {
+            base.OnInit(e);
+            ViewStateUserKey = Session.SessionID;
```
We have sessions disabled for our app, so this will fail. Not sure what this is needed for, but it will break our Admin :-)

scottbommarito replied:
It prevents CSRF. I'm not sure what Admin does so I'll have to ask you a question about it, but we have to set this to something or else we're vulnerable to attacks.

maartenba replied:
Can we set it to another value? Session state will fail.

scottbommarito replied:
Definitely, it just has to be something specific to each user, like a username (https://github.com/NuGet/NuGetGallery/pull/3173#issuecomment-236056228)

maartenba replied:
@blowdart would the logged in username work? Or a nameidentifier?

maartenba commented on src/NuGetGallery/NuGetGallery.csproj:
```diff
@@ -48,7 +48,7 @@
     <WarningLevel>4</WarningLevel>
     <Prefer32Bit>false</Prefer32Bit>
     <UseVSHostingProcess>true</UseVSHostingProcess>
-    <CodeAnalysisRuleSet>..\Frontend.ruleset</CodeAnalysisRuleSet>
```
@scottbommarito If the SDL plugin is not installed, will the build fail because the rule names can not be evaluated? (thinking about people building their own copy of the gallery)

If not, then ignore the rest of my comment.

If the build fails in that case: can we make this ruleset conditional? E.g. use `Frontend.ruleset` by default, and use the SDL one when a specific msbuild/environment property is set.

That property could be one we can set when on corpnet/when having the SDL ruleset installed (pick a name and document it). TeamCity has the `TEAMCITY_VERSION` environment variable. If that one exists, use the SDL-enabled ruleset.

(also applies to the next few ruleset inclusions in this PR)

scottbommarito replied:
Code analysis does not run on build, so nobody will be affected unless they specifically go and attempt to run code analysis. I don't know if code analysis fails if the ruleset doesn't exist. I can look into the `TEAMCITY_VERSION` solution.

maartenba replied:
If it does not break anyone who does not have the plugin, we can safely keep as-is :-)

maartenba commented on src/NuGetGallery/Web.config:
```diff
@@ -232,6 +232,10 @@
       </system.Web>
   -->
   <system.web>
+    <authentication mode="Forms">
```
Make sure to test this locally in the browser.

scottbommarito replied:
Just did, had to change the loginUrl below for it to work

maartenba replied:
🎉

maartenba commented on src/BackendSDL.ruleset:
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
UTF8 lead bytes? Maybe remove them...

scottbommarito replied:
Visual Studio generated this file from a wizard, I'll change it tho.

scottbommarito replied:
Actually the pre-existing `.ruleset` files (`Frontend.ruleset` and `Backend.ruleset`) have this line as well. What are you suggesting we change it to?

scottbommarito replied:
@shishirx34 explained it to me. It's rather peculiar that you can't see the leading bytes in Conversation on GitHub, but you can see them in the Files Changed!

maartenba replied:
I saw some weird characters at the beginning, but not anymore... Diff viewer, sigh.

scottbommarito replied:
Yeah it's SO WEIRD. I tried looking for them in VS and VS Code and even Notepad but I could never find and delete them! So spooky.

maartenba commented on src/FrontendSDL.ruleset:
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
UTF8 lead bytes? Maybe remove them...

skofman1 commented:
> ��[ a s s e m b l y :   S y s t e m . D i a g n o s t i c s . C o d e A n a l y s i s . S u p p r e s s M e s s a g e ( " M i c r o s o f t . S e c u r i t y . W e b . C o n f i g u r a t i o n " ,   " C A 3 1 0 3 : E n a b l e F o r m s R e q u i r e S S L " ,   M e s s a g e I d   =   " v i e w s / w e b . c o n f i g # c o n f i g u r a t i o n / s y s t e m . w e b / a u t h e n t i c a t i o n [ @ m o d e ] / f o r m s [ @ r e q u i r e S S L ] " ) ] 

There is something wrong with this file... Github won't show it in the PR and code flow shows it in a weird format.. make sure no strangeness happened..

---
Refers to: src/NuGetGallery/GlobalSuppressions.cs:1 in df2598e. [](commit_id = df2598e357415d449ab6e1c23312422b954b1885, deletion_comment = False)

scottbommarito commented:
@skofman1 That's really weird...it looks the same as it did before I pushed. I think it's because it's all just tags and no actual lines of code so GitHub got confused.

scottbommarito commented on src/NuGetGallery/Areas/Admin/DynamicData/Default.aspx.cs:
```diff
@@ -8,6 +8,12 @@ namespace NuGetGallery.Areas.Admin.DynamicData
+        protected override void OnInit(EventArgs e)
+        {
+            base.OnInit(e);
+            ViewStateUserKey = User.Identity.Name;
```
@blowdart Does this look right?

blowdart commented:
Works for me :)

scottbommarito commented:
@maartenba @skofman1 @shishirx34 @ryuyu 
Everything is resolved now! I just finished adding HTTP verb restrictions to every route, corrected the web forms authentication, and fixed an authentication issue with the Contact Us form.

maartenba commented:
LGTM, although I would like to see someone buddy-test these changes locally first before merging :-) (will give it a stab myself on monday, but an additional someone wouldn't hurt).

Also can you try squashing commits?

scottbommarito commented:
@maartenba Tbh I think I've tried squashing commits on this twice or three times (each time in the commit history it says "merge branch") but it doesn't seem to be actually doing anything...

blowdart commented:
Please remember the SDL rulesets are internal, and can't be pushed into source control.

scottbommarito commented:
@blowdart This isn't pushing the ruleset, it's pushing something that says "run this ruleset" to source control. Anybody who tries to run code analysis on this repo shouldn't have any success unless they have the SDL ruleset already on their PC. Correct me if I'm wrong, but this should be fine right?

blowdart commented:
Yup that's fine, and of course your build server runs the rules on build :)

scottbommarito commented:
@skofman1 @ryuyu @maartenba 
Need one more!

skofman1 commented on src/NuGetGallery/Areas/Admin/DynamicData/PageTemplates/ListDetails.aspx.cs:
```diff
@@ -20,7 +20,7 @@ protected void Page_Init(object sender, EventArgs e)
             GridDataSource.EntityTypeFilter = table.EntityType.Name;
             DetailsDataSource.EntityTypeFilter = table.EntityType.Name;
-
+            ViewStateUserKey = User.Identity.Name;
```
>ViewStateUserKey [](start = 12, length = 16)

this repeats a lot. Can we put it in a common location?

scottbommarito replied:
I don't think so, I believe it needs to be on every .aspx.cs in the same way that every .cshtml needs an Html.AntiForgeryToken on it...

maartenba replied:
Wouldn't bother too much, these templates come from ASP.NET Dynamic Data and are sort of generated. (package hasnt gotten any new updates since ages so no regeneration will happen anythime soon)

maartenba commented:
After the squash, do a `git push --force`. Never do this on the dev/master branch though as it essentially will break your colleagues :-) Only git push force on own branches...

scottbommarito commented:
@maartenba done! thanks

scottbommarito commented:
@skofman1 @ryuyu @shishirx34 
Can somebody review this so we can get the security issues fixed asap?

shishirx34 commented:
LGTM
